aggregators/merger.go

```diff
@@ -395,8 +397,32 @@ func mergeHistogram(to, from *aggregationpb.HDRHistogram) {
 		return
 	}
 
-	requiredLen := len(to.Buckets) + len(from.Buckets)
-	for toIdx, fromIdx := 0, 0; toIdx < len(to.Buckets) && fromIdx < len(from.Buckets); {
+	toLen, fromLen := len(to.Buckets), len(from.Buckets)
+	if fromLen == 1 {
+		// Fast path to optimize for cases where len(from.Buckets) == 1
+		// Binary search for from.Buckets[0] in to.Buckets,
+		// mergeHistogram will be O(lg N) where N = len(to.Buckets).
+		fromIdx := 0
+		fromVal := from.Buckets[fromIdx]
+		toIdx, found := sort.Find(toLen, func(toIdx int) int {
+			return int(fromVal - to.Buckets[toIdx])
+		})
+		if found {
+			to.Counts[toIdx] += from.Counts[fromIdx]
+		} else {
+			to.Buckets = slices.Grow(to.Buckets, 1)[:toLen+1]
+			to.Counts = slices.Grow(to.Counts, 1)[:toLen+1]
+			// Right shift by 1 position, then insert.
+			copy(to.Buckets[toIdx+1:toLen+1], to.Buckets[toIdx:toLen])
+			copy(to.Counts[toIdx+1:toLen+1], to.Counts[toIdx:toLen])
+			to.Buckets[toIdx] = from.Buckets[fromIdx]
+			to.Counts[toIdx] = from.Counts[fromIdx]
+		}
+		return
+	}
```

Comment:
I think we can make it simpler by following this algorithm:

1. For the first entry in `from` find the index in `to` which is just greater or equal
2. Reslice to to that index
3. Let the old logic do its job

This should cover the optimization for 1 value as well. WDYT?

Reply:
I can try to do that but we'll also need step 4 to reslice to its original range to offset shorter reslice in step 2

Reply:
Actually, since restoring to.Buckets and to.Counts at the end of function also means that we'll need to store a reference to the original slice, I suspect that it may look more complicated than it should. It seems that a dedicated fast path with `return` will be easier to understand. Step 1 also doesn't bring much speed up to cases where toLen is in the same order as fromLen, so I'm not sure if it is worth it to complicate the code for that.

Reply:
I've just pushed another commit to improve the slow path merge.

Reply:
Apologies for not explaining myself well, I meant something like below:

```go
func mergeHistogram(to, from *aggregationpb.HDRHistogram) {
	if len(from.Buckets) == 0 {
		return
	}

	if len(to.Buckets) == 0 {
		to.Buckets = append(to.Buckets, from.Buckets...)
		to.Counts = append(to.Counts, from.Counts...)
		return
	}

	startToIdx, found := sort.Find(len(to.Buckets), func(i int) int {
		return int(from.Buckets[0] - to.Buckets[i])
	})
	if found && len(from.Buckets) == 1 {
		// optimize for single value of `from` also found in `to`
		to.Counts[startToIdx] += from.Counts[0]
		return
	}

	// Since all values of `from` must be greater than the first value, we can limit the search
	// space in to to [startToIdx, len(to.Buckets))
	requiredLen := len(to.Buckets) + len(from.Buckets)
	for toIdx, fromIdx := startToIdx, 0; toIdx < len(to.Buckets) && fromIdx < len(from.Buckets); {
		v := to.Buckets[toIdx] - from.Buckets[fromIdx]
		switch {
		case v == 0:
			// For every bucket that is common, we need one less bucket in final slice
			requiredLen--
			toIdx++
			fromIdx++
		case v < 0:
			toIdx++
		case v > 0:
			fromIdx++
		}
	}

	toIdx, fromIdx := len(to.Buckets)-1, len(from.Buckets)-1
	to.Buckets = slices.Grow(to.Buckets, requiredLen-len(to.Buckets))
	to.Counts = slices.Grow(to.Counts, requiredLen-len(to.Counts))
	to.Buckets = to.Buckets[:requiredLen]
	to.Counts = to.Counts[:requiredLen]
	for idx := len(to.Buckets) - 1; idx >= 0; idx-- {
		if fromIdx < 0 {
			break
		}
		if toIdx < startToIdx {
			to.Counts[idx] = from.Counts[fromIdx]
			to.Buckets[idx] = from.Buckets[fromIdx]
			fromIdx--
			continue
		}
		v := to.Buckets[toIdx] - from.Buckets[fromIdx]
		switch {
		case v == 0:
			to.Counts[toIdx] += from.Counts[fromIdx]
			to.Counts[idx] = to.Counts[toIdx]
			to.Buckets[idx] = to.Buckets[toIdx]
			toIdx--
			fromIdx--
		case v > 0:
			to.Counts[idx] = to.Counts[toIdx]
			to.Buckets[idx] = to.Buckets[toIdx]
			toIdx--
		case v < 0:
			to.Counts[idx] = from.Counts[fromIdx]
			to.Buckets[idx] = from.Buckets[fromIdx]
			fromIdx--
		}
	}
}
```